### ISSUE TITLE
TestPerformanceIntegration.test_throughput failure

### ISSUE THREAD

**Issue by twm** [ciena/afkak#36]
I have only seen this [once](https://travis-ci.org/ciena/afkak/jobs/422850671):

```
ERROR: afkak.test.test_performance_integration.TestPerformanceIntegration.test_throughput
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/ciena/afkak/.tox/py35-int-snappy/lib/python3.5/site-packages/twisted/internet/defer.py", line 151, in maybeDeferred
    result = f(*args, **kw)
  File "/home/travis/build/ciena/afkak/.tox/py35-int-snappy/lib/python3.5/site-packages/twisted/internet/utils.py", line 201, in runWithWarningsSuppressed
    reraise(exc_info[1], exc_info[2])
  File "/home/travis/build/ciena/afkak/.tox/py35-int-snappy/lib/python3.5/site-packages/twisted/python/compat.py", line 463, in reraise
    raise exception.with_traceback(traceback)
  File "/home/travis/build/ciena/afkak/.tox/py35-int-snappy/lib/python3.5/site-packages/twisted/internet/utils.py", line 197, in runWithWarningsSuppressed
    result = f(*a, **kw)
  File "/home/travis/build/ciena/afkak/.tox/py35-int-snappy/lib/python3.5/site-packages/nose/twistedtools.py", line 169, in wrapper
    raise exc_type(exc_value).with_traceback(tb)
  File "/home/travis/build/ciena/afkak/.tox/py35-int-snappy/lib/python3.5/site-packages/nose/twistedtools.py", line 143, in errback
    failure.raiseException()
  File "/home/travis/build/ciena/afkak/.tox/py35-int-snappy/lib/python3.5/site-packages/twisted/python/failure.py", line 467, in raiseException
    raise self.value.with_traceback(self.tb)
  File "/home/travis/build/ciena/afkak/.tox/py35-int-snappy/lib/python3.5/site-packages/twisted/internet/defer.py", line 1418, in _inlineCallbacks
    result = g.send(result)
  File "/home/travis/build/ciena/afkak/afkak/test/testutil.py", line 211, in tearDown
    self.assertFalse(dcs)
  File "/home/travis/build/ciena/afkak/.tox/py35-int-snappy/lib/python3.5/site-packages/twisted/trial/_synctest.py", line 384, in assertFalse
    super(_Assertions, self).assertFalse(condition, msg)
  File "/opt/python/3.5.6/lib/python3.5/unittest/case.py", line 676, in assertFalse
    raise self.failureException(msg)
twisted.trial.unittest.FailTest: [<twisted.internet.base.DelayedCall object at 0x7f9b3a024c18>] is not false
```

**Comment by rthille** [comment]
Something isn't properly waiting on a deferred. It's been "lost".

**Comment by twm** [comment]
https://travis-ci.org/ciena/afkak/jobs/432714302

**Comment by rthille** [comment]
Kinda looks like the logging got cut off. Maybe it's due to not handling a failure during the test properly?

**Comment by twm** [comment]
I suspect this is actually something left in the reactor by a previous test. I _never_ see it when running just `test_throughput`, but it happens reliably on my desktop when running the whole suite. Furthermore, [in this failure](https://travis-ci.org/ciena/afkak/jobs/433847467#L686) there is a `ConsumerGroup` delayed call, when consumer groups are not used in the throughput test:

```
======================================================================
ERROR: afkak.test.test_performance_integration.TestPerformanceIntegration.test_throughput
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/ciena/afkak/.tox/py27-int-snappy-murmur/lib/python2.7/site-packages/twisted/internet/defer.py", line 150, in maybeDeferred
    result = f(*args, **kw)
  File "/home/travis/build/ciena/afkak/.tox/py27-int-snappy-murmur/lib/python2.7/site-packages/twisted/internet/utils.py", line 201, in runWithWarningsSuppressed
    reraise(exc_info[1], exc_info[2])
  File "/home/travis/build/ciena/afkak/.tox/py27-int-snappy-murmur/lib/python2.7/site-packages/twisted/internet/utils.py", line 197, in runWithWarningsSuppressed
    result = f(*a, **kw)
  File "/home/travis/build/ciena/afkak/.tox/py27-int-snappy-murmur/lib/python2.7/site-packages/nose/twistedtools.py", line 143, in errback
    failure.raiseException()
  File "/home/travis/build/ciena/afkak/.tox/py27-int-snappy-murmur/lib/python2.7/site-packages/twisted/internet/defer.py", line 1386, in _inlineCallbacks
    result = g.send(result)
  File "/home/travis/build/ciena/afkak/afkak/test/testutil.py", line 233, in tearDown
    self.assertNoDelayedCalls()
  File "/home/travis/build/ciena/afkak/afkak/test/testutil.py", line 197, in assertNoDelayedCalls
    '\n'.join(str(dc) for dc in dcs),
AssertionError: Found 2 outstanding delayed calls:
<DelayedCall 0x7fcd48e8d248 [0.258409976959s] called=0 cancelled=0 ConsumerGroup.join_and_sync()>
<DelayedCall 0x7fcd48fd87a0 [13.7548611164s] called=0 cancelled=0 reconnector()>
```

**Comment by rthille** [comment]
Hmm, surprised the previous test's tearDown didn't catch it with that assert then.

**Comment by twm** [comment]
Yeah, there definitely is such an assertion. However when it trips it doesn't clean up the reactor, so future tests are poisoned. I tried adding a delayed call to the test Nose runs right before test_throughput and got these results:

```
#33 afkak.test.test_producer_integration.TestAfkakProducerIntegration.test_producer_send_messages ... ok (1.5389s)
#34 afkak.test.test_producer_integration.TestAfkakProducerIntegration.test_write_nonextant_topic ... ok (7.9128s)

======================================================================
ERROR: test suite for <class 'afkak.test.test_failover_integration.TestFailover'>
----------------------------------------------------------------------
Traceback (most recent call last):
  File ".../afkak/.tox/py35-int-snappy/lib/python3.5/site-packages/nose/suite.py", line 229, in run
    self.tearDown()
  File ".../afkak/.tox/py35-int-snappy/lib/python3.5/site-packages/nose/suite.py", line 352, in tearDown
    self.teardownContext(ancestor)
  File ".../afkak/.tox/py35-int-snappy/lib/python3.5/site-packages/nose/suite.py", line 368, in teardownContext
    try_run(context, names)
  File ".../afkak/.tox/py35-int-snappy/lib/python3.5/site-packages/nose/util.py", line 471, in try_run
    return func()
  File ".../afkak/.tox/py35-int-snappy/lib/python3.5/site-packages/nose/twistedtools.py", line 169, in wrapper
    raise exc_type(exc_value).with_traceback(tb)
  File ".../afkak/.tox/py35-int-snappy/lib/python3.5/site-packages/nose/twistedtools.py", line 143, in errback
    failure.raiseException()
  File ".../afkak/.tox/py35-int-snappy/lib/python3.5/site-packages/twisted/python/failure.py", line 385, in raiseException
    raise self.value.with_traceback(self.tb)
  File ".../afkak/.tox/py35-int-snappy/lib/python3.5/site-packages/twisted/internet/defer.py", line 1386, in _inlineCallbacks
    result = g.send(result)
  File ".../afkak/afkak/test/test_failover_integration.py", line 68, in tearDownClass
    cls.assertNoDelayedCalls()
  File ".../afkak/afkak/test/testutil.py", line 197, in assertNoDelayedCalls
    '\n'.join(str(dc) for dc in dcs),
AssertionError: Found 1 outstanding delayed calls:
<DelayedCall 0x7f566c443240 [59.99595642089844s] called=0 cancelled=0 TestFailover.test_switch_leader.<locals>.<lambda>()>

======================================================================
ERROR: afkak.test.test_performance_integration.TestPerformanceIntegration.test_throughput
----------------------------------------------------------------------
Traceback (most recent call last):
  File ".../afkak/.tox/py35-int-snappy/lib/python3.5/site-packages/twisted/internet/defer.py", line 150, in maybeDeferred
    result = f(*args, **kw)
  File ".../afkak/.tox/py35-int-snappy/lib/python3.5/site-packages/twisted/internet/utils.py", line 201, in runWithWarningsSuppressed
    reraise(exc_info[1], exc_info[2])
  File ".../afkak/.tox/py35-int-snappy/lib/python3.5/site-packages/twisted/python/compat.py", line 463, in reraise
    raise exception.with_traceback(traceback)
  File ".../afkak/.tox/py35-int-snappy/lib/python3.5/site-packages/twisted/internet/utils.py", line 197, in runWithWarningsSuppressed
    result = f(*a, **kw)
  File ".../afkak/.tox/py35-int-snappy/lib/python3.5/site-packages/nose/twistedtools.py", line 169, in wrapper
    raise exc_type(exc_value).with_traceback(tb)
  File ".../afkak/.tox/py35-int-snappy/lib/python3.5/site-packages/nose/twistedtools.py", line 143, in errback
    failure.raiseException()
  File ".../afkak/.tox/py35-int-snappy/lib/python3.5/site-packages/twisted/python/failure.py", line 385, in raiseException
    raise self.value.with_traceback(self.tb)
  File ".../afkak/.tox/py35-int-snappy/lib/python3.5/site-packages/twisted/internet/defer.py", line 1386, in _inlineCallbacks
    result = g.send(result)
  File ".../afkak/afkak/test/testutil.py", line 233, in tearDown
    self.assertNoDelayedCalls()
  File ".../afkak/afkak/test/testutil.py", line 197, in assertNoDelayedCalls
    '\n'.join(str(dc) for dc in dcs),
AssertionError: Found 1 outstanding delayed calls:
<DelayedCall 0x7f566c443240 [30.49918031692505s] called=0 cancelled=0 TestFailover.test_switch_leader.<locals>.<lambda>()>

======================================================================
ERROR: afkak.test.test_performance_integration.TestPerformanceIntegration.test_throughput
----------------------------------------------------------------------
twisted.trial.util.DirtyReactorAggregateError: Reactor was unclean.
DelayedCalls: (set twisted.internet.base.DelayedCall.debug = True to debug)
<DelayedCall 0x7f566c443240 [30.49620819091797s] called=0 cancelled=0 TestFailover.test_switch_leader.<locals>.<lambda>()>

======================================================================
FAIL: afkak.test.test_failover_integration.TestFailover.test_switch_leader
----------------------------------------------------------------------
Traceback (most recent call last):
  File ".../afkak/.tox/py35-int-snappy/lib/python3.5/site-packages/nose/twistedtools.py", line 169, in wrapper
    raise exc_type(exc_value).with_traceback(tb)
  File ".../afkak/.tox/py35-int-snappy/lib/python3.5/site-packages/nose/twistedtools.py", line 143, in errback
    failure.raiseException()
  File ".../afkak/.tox/py35-int-snappy/lib/python3.5/site-packages/twisted/python/failure.py", line 385, in raiseException
    raise self.value.with_traceback(self.tb)
AssertionError: Found 1 outstanding delayed calls:
<DelayedCall 0x7f566c443240 [59.99902868270874s] called=0 cancelled=0 TestFailover.test_switch_leader.<locals>.<lambda>()>

<nose.suite.ContextSuite context=TestFailover>:teardown: 30.9159s
afkak.test.test_failover_integration.TestFailover.test_switch_leader: 30.9129s
afkak.test.test_performance_integration.TestPerformanceIntegration.test_throughput: 25.8774s
afkak.test.test_producer_integration.TestAfkakProducerIntegration.test_write_nonextant_topic: 7.9128s
afkak.test.test_producer_integration.TestAfkakProducerIntegration.test_producer_batched_by_time: 6.0725s
afkak.test.test_producer_integration.TestAfkakProducerIntegration.test_producer_batched_gzipped_hashed_partitioner: 6.0399s
afkak.test.test_consumer_integration.TestConsumerIntegration.test_consumer: 3.7195s
afkak.test.test_producer_integration.TestAfkakProducerIntegration.test_producer_batched_by_bytes: 3.5484s
afkak.test.test_producer_integration.TestAfkakProducerIntegration.test_producer_batched_by_messages: 3.0434s
afkak.test.test_consumer_integration.TestConsumerIntegration.test_consumer_commit_offsets: 2.4684s
----------------------------------------------------------------------
Ran 33 tests in 195.067s

FAILED (errors=3, failures=1)
```

So I think this indicates that this particular failure is a side effect of other failures. The underlying problem is either a bug in the test (not shutting down the client fully) or the client (not cancelling deferred calls when shut down).

**Comment by rthille** [comment]
Yeah, I don't think the tests did a great job of cleaning up if they didn't pass.  If they failed early, they likely leave trash around. Sorry :-(

**Comment by twm** [comment]
https://travis-ci.org/ciena/afkak/jobs/442955282

**Comment by twm** [comment]
https://travis-ci.org/ciena/afkak/jobs/442955283

**Comment by twm** [comment]
The rogue delayed call originates here: https://github.com/ciena/afkak/blob/8cf2301e33a57ef6f26b53c7ff4af8d5918e9f13/afkak/producer.py#L541-L543

**Comment by rthille** [comment]
I think you're right that that deferred isn't kept (directly) in any Producer data structure, but it's chained off the `self._batch_send_d` deferred, and should get cancelled when that gets cancelled.
Everything that needs to go on when starting to process the batch *should* be chained off that `self._batch_send_d` deferred.
It's a horribly complicated mess though.

**Comment by twm** [comment]
Here is a cleaner traceback:

```
ERROR: afkak.test.test_performance_integration.TestPerformanceIntegration.test_throughput
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/ciena/afkak/.tox/py27-int-snappy-murmur/lib/python2.7/site-packages/twisted/internet/defer.py", line 151, in maybeDeferred
    result = f(*args, **kw)
  File "/home/travis/build/ciena/afkak/.tox/py27-int-snappy-murmur/lib/python2.7/site-packages/twisted/internet/utils.py", line 201, in runWithWarningsSuppressed
    reraise(exc_info[1], exc_info[2])
  File "/home/travis/build/ciena/afkak/.tox/py27-int-snappy-murmur/lib/python2.7/site-packages/twisted/internet/utils.py", line 197, in runWithWarningsSuppressed
    result = f(*a, **kw)
  File "/home/travis/build/ciena/afkak/.tox/py27-int-snappy-murmur/lib/python2.7/site-packages/nose/twistedtools.py", line 143, in errback
    failure.raiseException()
  File "/home/travis/build/ciena/afkak/.tox/py27-int-snappy-murmur/lib/python2.7/site-packages/twisted/internet/defer.py", line 1418, in _inlineCallbacks
    result = g.send(result)
  File "/home/travis/build/ciena/afkak/afkak/test/testutil.py", line 227, in tearDown
    self.assertNoDelayedCalls()
  File "/home/travis/build/ciena/afkak/afkak/test/testutil.py", line 197, in assertNoDelayedCalls
    '\n'.join(str(dc) for dc in dcs),
AssertionError: Found 1 outstanding delayed calls:
<DelayedCall 0x7f0881ab5170 [0.211324930191s] called=0 cancelled=0 Deferred.callback([ProduceRequest(topic=u'test_throughput-LINZWEwkJz', partition=10, messages=[Message(magic=0, attributes=0, key=None, value='gtdqhsTLnOv...
```

It comes from just after the client is closed:

https://github.com/ciena/afkak/blob/8cf2301e33a57ef6f26b53c7ff4af8d5918e9f13/afkak/test/testutil.py#L219-L227

**Comment by rthille** [comment]
Yeah, that makes sense. But from code inspection it *seems* like that delayed call should get cancelled if that deferred it's associated with is cancelled, which *should* happen when producer.stop() is called.  But yeah, trying to understand that deferred chain that hangs off of producer's `self._batch_send_d` is seriously ugly, as every operation/retry/etc is supposed to hang off that one deferred, but there are a ton of things that can return a new deferred during processing to hang off it.
